File: data-structures/non-linear/binary-tree/recursive/BinaryTree.ts
import { TreeNode } from "../../node/TreeNode";

let index: number = -1;

function buildTree(preOrderArray: Array<number>): TreeNode | null {
  index++;

  if (index >= preOrderArray.length || preOrderArray[index] == -1) {
    return null;
  }

  let root: TreeNode = new TreeNode(preOrderArray[index]);
  root.setLeft(buildTree(preOrderArray));
  root.setRight(buildTree(preOrderArray));

  return root;
}

function preOrderTraversal(root: TreeNode | null): void {
  if (root == null) {
    return;
  }

  console.log(`${root.getValue()}`);
  preOrderTraversal(root.getLeft());
  preOrderTraversal(root.getRight());
}

function inOrderTraversal(root: TreeNode | null): void {
  if (root == null) {
    return;
  }

  inOrderTraversal(root.getLeft());
  console.log(`${root.getValue()}`);
  inOrderTraversal(root.getRight());
}

function postOrderTraversal(root: TreeNode | null): void {
  if (root == null) {
    return;
  }

  postOrderTraversal(root.getLeft());
  postOrderTraversal(root.getRight());
  console.log(`${root.getValue()}`);
}

function levelOrderTraversal(root: TreeNode | null): void {
  if (root == null) {
    return;
  }

  let queue: Array<TreeNode | null> = new Array();
  let temp: TreeNode | null | undefined; // as array.shift() can return undefined
  queue.push(root);
  queue.push(null);

  while (queue.length != 0) {
    temp = queue.shift();

    if (temp == undefined) {
      continue;
    }

    if (temp == null) {
      if (queue.length != 0) {
        queue.push(null);
      }
    } else {
      console.log(`${temp.getValue()}`);

      if (temp.getLeft() != null) {
        queue.push(temp.getLeft());
      }

      if (temp.getRight() != null) {
        queue.push(temp.getRight());
      }
    }
  }
}

export {
  buildTree,
  inOrderTraversal,
  levelOrderTraversal,
  postOrderTraversal,
  preOrderTraversal,
};

File: data-structures/non-linear/binary-tree/recursive/BinaryTree.java
import java.util.*;

class Node {
  int data;
  Node left;
  Node right;

  Node(int value) {
    data = value;
    left = right = null;
  }
}

public class BinaryTree {
  static int index = -1;

  public static Node buildTree(int[] preorder) {
    index++;

    if (index >= preorder.length || preorder[index] == -1) {
      return null;
    }

    Node root = new Node(preorder[index]);
    root.left = buildTree(preorder);
    root.right = buildTree(preorder);

    return root;
  }

  public static void preorderTraversal(Node root) {
    if (root == null)
      return;

    System.out.print(root.data + " ");
    preorderTraversal(root.left);
    preorderTraversal(root.right);
  }

  public static void inorderTraversal(Node root) {
    if (root == null)
      return;

    inorderTraversal(root.left);
    System.out.print(root.data + " ");
    inorderTraversal(root.right);
  }

  public static void postorderTraversal(Node root) {
    if (root == null)
      return;

    postorderTraversal(root.left);
    postorderTraversal(root.right);
    System.out.print(root.data + " ");
  }

  public static void levelorderTraversal(Node root) {
    if (root == null)
      return;

    Queue<Node> q = new LinkedList<>();
    q.add(root);
    q.add(null);

    while (!q.isEmpty()) {
      Node temp = q.poll();

      if (temp == null) {
        System.out.println();
        if (!q.isEmpty()) {
          q.add(null);
        }
      } else {
        System.out.print(temp.data + " ");

        if (temp.left != null) {
          q.add(temp.left);
        }

        if (temp.right != null) {
          q.add(temp.right);
        }
      }
    }
  }
}
